CRM/Volunteer/BAO/Project.php

```diff
@@ -269,13 +292,18 @@ public static function create(array $params) {
         // CRM-17222: For compatibility with CiviCRM 4.6, we use our custom API.
         // When support for 4.6 is dropped, we can use core's api.UFJoin.delete.
         // 'api.UFJoin.delete' => array(),
-        'api.VolunteerProject.removeprofile' => array(),
+        'api.VolunteerProject.removeprofile' => array(
+          // For some reason, the ID param implicit in chained API calls is not
+          // getting passed, so we specify it manually. Since the API is
+          // deprecated, this short-term solution is good enough.
+          'id' => '$value.id',
+        ),
       ));
     }
 
-    if ($updateVPC || CRM_Core_Action::ADD) {
+    if ($updateVPC || $op === CRM_Core_Action::ADD) {
```

Comment:
I think this is evidence of how code suffers when you mix concerns. This test should be "was project_contacts passed?" but you can't do that because supply_defaults pollutes the params.

I think this function is bending over backwards to implement implicit deletes for project-relationships and uf_joins. Moving the logic for ufjoins and project-relationships to separate functions would make this read much more easily. It might then be okay to not have distinct default functions which might be more transparent. A prototypical create function could be: check permissions; validate; default; ensure child-objects; persist.

I know I'm belabouring this function, but I feel like bad idioms in CiviCRM are infecting the code and I see how this could get ugly down the road as Project entities get more complex, or we add finer-grained permissions, etc. Dependant lines of code are getting separated and will get more separated as the code grows. Ideally, I prefer to see a logic-branch just once. Branching on the same test is a red-flag for maintainability and a good indicator of mixing concerns.